### PR TITLE
chore(sdk): topological ordering of `wandb.Settings`

### PR DESCRIPTION
**Pull request by dmitryduev**
Description
-----------
Individual settings may depend on other settings through validator methods and runtime hooks, but the resulting directed dependency graph must be acyclic. 
In this PR: 
- Automatic cyclic dependency detection. We build the dependency DAG by inspecting the validator methods and runtime hooks and topologically sort it.
- Ensure that modifications to the Settings object (or copying thereof) are performed in the order that respects the dependency DAG, i.e. follows the topological ordering.

Testing
-------
Added unit tests.

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -649,10 +653,11 @@ not perform any serialization to sync data.
...
 ```python
-1 import wandb
-2 run = wandb.init(config=dict(param1=1))
-3 run.config.param2 = 2
-4 run.log(dict(this=3))
+import wandb
+
+run = wandb.init(config=dict(param1=1))
+run.config.param2 = 2
+run.log(dict(this=3))
```
Comment (dmitryduev) [inline CONTRIBUTING.md:660]: doc-code linter now complains about this.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -778,7 +783,7 @@ To mark a feature as deprecated (and to be removed in the next major release), p
 from wandb.sdk.lib import deprecate
 
 deprecate.deprecate(
-    field_name=deprecate.Deprecated.<new_field_name>,  # new_field_name from step 1
+    field_name=deprecate.Deprecated.new_field_name,  # new_field_name from step 1
```
Comment (dmitryduev) [inline CONTRIBUTING.md:786]: and this

**Comment by dmitryduev** [comment]
@speezepearson this is the joyful PR I wanted to share with you, not the #4026 rubbish :)

File: wandb/sdk/wandb_settings.py [hunk 3]
```diff
@@ -155,6 +157,63 @@ def _get_program_relpath_from_gitrepo(
...
 
 
+class Graph:
+    # A simple class representing an unweighted directed graph
+    # that uses an adjacency list representation.
+    # We use to ensure that we don't have cyclic dependencies in the settings
+    # and that modifications to the settings are applied in the correct order.
+    def __init__(self) -> None:
+        self.adj_list: Dict[str, Set[str]] = {}
+        self.nodes: Set[str] = set()
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:167]: shy do you need nodes if you already have the `adj_list`?

File: wandb/sdk/wandb_settings.py [hunk 4]
```diff
@@ -155,6 +157,63 @@ def _get_program_relpath_from_gitrepo(
...
+            for neighbor in self.get_neighbors(n):
+                visit(neighbor)
+
+            current_nodes.remove(n)
+            visited_nodes.add(n)
+            sorted_nodes.append(n)
+
+            return None
+
+        for node in self.nodes:
```
Comment (kptkin) [inline wandb/sdk/wandb_settings.py:210]: should you use the `getter` here?

**Review by kptkin — APPROVED** [review]
🔥

**Review by raubitsj — APPROVED** [review]
NICE..
might want to add the generated file to .gitattributes

**Pull request merged**